### PR TITLE
fix: "filter dashboard to" for null value

### PR DESCRIPTION
**Pull request by joaoviana**
Closes: #8891

### Description:

On a dashboard, when the value on a cell or chart point is `null`, and when users clicked on `filter dashboard to` the following happened: 
* `filter dashboard to` with empty value
* when user clicked on that (to filter the dashboard by the `null` value), it wouldn't create a temp filter with `is null` operator, it would instead create a filter with `is/EQUAL` with empty value. 

This PR adds:
- seed data to the events table that includes `null`/empty `event_id`s
- utilises the util `createDashboardFilterRuleFromField` more across the codebase 
- creates a shared component `FilterDashboardTo` that is used on the char tiles and cell context menu

Screen recording:

https://github.com/lightdash/lightdash/assets/7611706/ec9f2b59-00f2-4be0-a22f-2757166de9cb

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx [hunk 1]
```diff
@@ -59,34 +58,36 @@ const DashboardCellContextMenu: FC<
...
+
+        acc.push(
+            createDashboardFilterRuleFromField({
+                field: pivotField,
+                availableTileFilters: {},
+                isTemporary: true,
+                value: pivot.value,
+            }),
+        );
+        return acc;
```
Comment (IrakliJani) [inline packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx:89]: nit: this can be `return [...acc, createDashboarblabla]` to avoid using push

File: packages/frontend/src/components/CustomSqlPanel/CustomMetricsTable.tsx [hunk 2]
```diff
@@ -67,11 +67,6 @@ const CustomMetricsTable: FC<Props> = ({
         [checkedUuids],
     );
     const toggleAllCheckedUuids = useCallback(() => {
-        console.log(
-            'toggleAllCheckedUuids',
-            checkedUuids.size,
-            customMetrics.length,
-        );
```
Comment (IrakliJani) [inline packages/frontend/src/components/CustomSqlPanel/CustomMetricsTable.tsx:74]: logs

File: packages/frontend/src/components/DashboardTiles/DashboardChartTile.tsx [hunk 3]
```diff
@@ -750,40 +745,10 @@ const DashboardChartTileMain: FC<DashboardChartTileMainProps> = (props) => {
...
-                                                            filter.values[0],
-                                                        )}
-                                                </Text>
-                                            </Menu.Item>
-                                        ),
-                                    )}
-                                </>
+                                <FilterDashboardTo
+                                    filters={dashboardTileFilterOptions}
+                                    addFilter={handleAddFilter}
```
Comment (IrakliJani) [inline packages/frontend/src/components/DashboardTiles/DashboardChartTile.tsx:750]: nit: `onAddFilter={handleAddFilter}`

File: packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx [hunk 4]
```diff
@@ -59,34 +58,36 @@ const DashboardCellContextMenu: FC<
...
-            operator: FilterOperator.EQUALS,
-            values: [pivot.value],
-            label: undefined,
-        };
-    });
+        if (
+            !pivotField ||
+            !isField(pivotField) ||
+            !isFilterableField(pivotField)
+        )
```
Comment (IrakliJani) [inline packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx:78]: can you explain what is the reason behind this change?

File: packages/common/src/utils/filters.ts [hunk 5]
```diff
@@ -323,7 +329,7 @@ export const createDashboardFilterRuleFromField = (
             disabled: !isTemporary,
             label: undefined,
         },
-        includeDefaultValue ? [] : null,
+        value !== undefined ? [value] : [],
```
Comment (IrakliJani) [inline packages/common/src/utils/filters.ts:332]: so if we are creating a "is null" operator do we still have to provide it as a value?

operator: NULL
value: [null] << ?

File: packages/common/src/utils/filters.ts [hunk 6]
```diff
@@ -304,17 +304,23 @@ export const applyDefaultTileTargets = (
...
+    isTemporary: boolean;
+    value?: any;
+}): DashboardFilterRule =>
     getFilterRuleWithDefaultValue(
         field,
         {
             id: uuidv4(),
-            operator: FilterOperator.EQUALS,
+            operator:
+                value === null ? FilterOperator.NULL : FilterOperator.EQUALS,
```
Comment (IrakliJani) [inline packages/common/src/utils/filters.ts:323]: is this function being used somewhere where user might not create a filter like this one:
field: `name` (string)
operator: `INCLUDE`
values [null, 'irakli'] << not sure if this is possible today. just asking

File: packages/frontend/src/components/DashboardTiles/DashboardChartTile.tsx [hunk 7]
```diff
@@ -393,9 +392,8 @@ const DashboardChartTileMain: FC<DashboardChartTileMainProps> = (props) => {
         [],
     );
 
-    const [dashboardTileFilterOptions, setDashboardFilterOptions] = useState<
-        DashboardFilterRule[]
-    >([]);
+    const [dashboardTileFilterOptions, setDashboardTileFilterOptions] =
```
Comment (IrakliJani) [inline packages/frontend/src/components/DashboardTiles/DashboardChartTile.tsx:395]: 👍

**Review by IrakliJani — COMMENTED** [review]
looks good, nice tidy-up! have some Qs and nits

File: packages/frontend/src/components/DashboardTiles/DashboardChartTile.tsx (hunk 3 above)
Comment (joaoviana) [reply packages/frontend/src/components/DashboardTiles/DashboardChartTile.tsx:750]: nice!

File: packages/frontend/src/components/DashboardFilter/FilterDashboardTo.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    if (filter.values)
+        return (
+            <Text span fw={500}>
+                {String(filter.values[0])}
+            </Text>
+        );
+
+    return null;
+};
```
[the remark this answers is not in the document]
Comment (joaoviana) [reply packages/frontend/src/components/DashboardFilter/FilterDashboardTo.tsx:32]: sure, will move it down!

File: packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx (hunk 4 above)
Comment (joaoviana) [reply packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx:78]: oh because the `pivotField` has to be filterable (of this type Field | Dimension | Metric | TableCalculationField) 

When creating a Dashboard Filter Rule it has to be `FilterableField`, otherwise we don't allow it. This will work as before, just more type robust

File: packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx (hunk 1 above)
Comment (joaoviana) [reply packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx:89]: sure!

File: packages/common/src/utils/filters.ts (hunk 5 above)
Comment (joaoviana) [reply packages/common/src/utils/filters.ts:332]: Good spot, I mean `value === null`, thank you!

File: packages/common/src/utils/filters.ts (hunk 6 above)
Comment (joaoviana) [reply packages/common/src/utils/filters.ts:323]: thanks for spotting this! It makes sense to rework this a bit more to avoid the bug you've just mentioned.

File: packages/frontend/src/components/DashboardFilter/FilterConfiguration/index.tsx [hunk 9]
```diff
@@ -109,12 +109,11 @@ const FilterConfiguration: FC<Props> = ({
 
         if (newField && isField(newField) && isFilterableField(newField)) {
             setDraftFilterRule(
-                createDashboardFilterRuleFromField(
-                    newField,
+                createDashboardFilterRuleFromField({
+                    field: newField,
```
Comment (joaoviana) [inline packages/frontend/src/components/DashboardFilter/FilterConfiguration/index.tsx:113]: this is where it could be _wrong_ but it's not! 
At the moment, if we don't pass a `value` to this functions' object, we set the operator by default to `EQUALS` - this was done before 
ONly if we pass a `value` `null` to it we change its operator to `NULL`

File: packages/common/src/utils/filters.ts [hunk 10]
```diff
@@ -329,7 +329,7 @@ export const createDashboardFilterRuleFromField = ({
             disabled: !isTemporary,
             label: undefined,
         },
-        value === null ? undefined : [value],
+        value ? [value] : null, // When `null`, don't set default value if no value is provided
```
Comment (joaoviana) [inline packages/common/src/utils/filters.ts:332]: If value is `falsy`, don't set the default value that is handled by `getFilterRuleWithDefaultValue`

This works OK when: 
`filtering dashboard to...` and when `creating a filter`

File: packages/common/src/utils/filters.ts (hunk 6 above)
Comment (IrakliJani) [reply packages/common/src/utils/filters.ts:323]: thanks for the explanation. that makes sense

File: packages/common/src/utils/filters.ts [hunk 11]
```diff
@@ -304,17 +304,23 @@ export const applyDefaultTileTargets = (
...
+export const createDashboardFilterRuleFromField = ({
+    field,
+    availableTileFilters,
+    isTemporary,
+    value,
+}: {
+    field: FilterableField;
+    availableTileFilters: Record<string, FilterableField[] | undefined>;
+    isTemporary: boolean;
+    value?: any;
```
Comment (IrakliJani) [inline packages/common/src/utils/filters.ts:316]: I suggest using `unknown` which is "more safe"
```suggestion
    value?: unknown;
```

File: packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx (hunk 4 above)
Comment (IrakliJani) [reply packages/frontend/src/components/SimpleTable/DashboardCellContextMenu.tsx:78]: thanks!

File: packages/common/src/utils/filters.ts (hunk 10 above)
Comment (IrakliJani) [reply packages/common/src/utils/filters.ts:332]: so in this case if value is `null` then we don't need to include in the values array. code is a bit confusing but comment makes it clear. thanks

**Review by IrakliJani — COMMENTED** [review]
code looks good 🎉 
I'm going to test on render now before approving

**Review by IrakliJani — CHANGES_REQUESTED** [review]
@joaoviana I've noticed a weird behavior when working with "any value" filters

here's a video on how to reproduce it: https://github.com/lightdash/lightdash/assets/962095/8ace6cd5-e5b8-4557-aeaa-cc8882249618

File: packages/common/src/utils/filters.ts (hunk 11 above)
Comment (joaoviana) [reply packages/common/src/utils/filters.ts:316]: done!

**Comment by joaoviana** [comment]
> @joaoviana I've noticed a weird behavior when working with "any value" filters
> 
> here's a video on how to reproduce it: https://github.com/lightdash/lightdash/assets/962095/8ace6cd5-e5b8-4557-aeaa-cc8882249618

Replace the `any` to `unknown`

I've raised the bug you've reported here: https://github.com/lightdash/lightdash/issues/9029

The util that was refactored doesn't get called when changing the operator, so this is unrelated and is in `main` already 😢

**Review by IrakliJani — APPROVED** [review]
